### DISCUSSION TITLE
Is it possible to enable autoscaling on existing cluster?

### DISCUSSION THREAD

**Discussion by otavio** [mysticaltech/terraform-hcloud-kube-hetzner#372] [Q&A] [answered]
Considering an existing cluster is deployed, is it possible to add autoscaling support?

**Comment by mysticaltech** [comment, marked as answer]
Yes, normally you just upgrade the module and apply.

**Reply by mysticaltech** [reply]
With your autoscaler definitions added.

**Reply by mysticaltech** [reply]
In other words, 1.6 should be pretty much backward compatible with 1.5.x, however, we have not tested that use case, but please do try it and let us know please!

**Reply by ifeulner** [reply]
Yes, should work. Just ensure that the server_type for the pool specified is equal or better regarding disk size as your first control_plane node.

**Reply by Manikanta-Reddy-Pasala** [reply]
@otavio 

if you have done it please let me know the steps.

I have tried the below steps but looks like the autoscaling of machines not happening.

1) updated terraform file :
  autoscaler_nodepools = [
     {
       name        = "autoscaler"
       server_type = "cpx21" # must be same or better than the control_plane server type (regarding disk size)!
       location    = "nbg1"
       min_nodes   = 0
       max_nodes   = 3
     }
 ]

2) executed below commands:
terraform init --upgrade
terraform validate
terraform apply -auto-approve

I saw it created a snapshot

3) drained one node to apply high load in 2 nodes, then due to heavy pod load other node went down but even after new servers spinoff didn't happen.

Please let me know if my expectation is wrong.

**Reply by ifeulner** [reply]
Hi @Manikanta-Reddy-Pasala  what do you mean with "drained one node to apply high load"? Which node drained? Draining means a node gets deleted.
Scaling depends heavily of the load in your cluster. And of the individual pod requirements (cpu, memory). 
 
Also If there is not enough load in your cluster, no additional node will be scaled-up by the cluster autoscaler. 

Can you provide the status of the cluster autoscaler via `kubectl describe configmap cluster-autoscaler-status -n kube-system`?

**Reply by Manikanta-Reddy-Pasala** [reply]
I have 3 node cluster. one node drained so that all pods in node 3 will be moved to node1 and node2. and I know that (node1 and node 2) it can't handle the number of pods. i except since I have enabled autoscaling it will spinoff a new node to my cluster but that didn't happen.

then after a couple of minutes, node1 went to non-scheduled mode with high CPU use making my cluster down.

i have posted output from the command
**% kubectl describe configmap cluster-autoscaler-status -n kube-system**

Name:         cluster-autoscaler-status
Namespace:    kube-system
Labels:       <none>
Annotations:  cluster-autoscaler.kubernetes.io/last-updated: 2022-10-26 06:28:44.926359783 +0000 UTC

Data
====
status:
----
Cluster-autoscaler status at 2022-10-26 06:28:44.926359783 +0000 UTC:
Cluster-wide:
  Health:      Healthy (ready=4 unready=0 (resourceUnready=0) notStarted=0 longNotStarted=0 registered=4 longUnregistered=0)
               LastProbeTime:      2022-10-26 06:28:44.925215661 +0000 UTC m=+17956.713203768
               LastTransitionTime: 2022-10-26 01:29:58.652331458 +0000 UTC m=+30.440319565
  ScaleUp:     NoActivity (ready=4 registered=4)
               LastProbeTime:      2022-10-26 06:28:44.925215661 +0000 UTC m=+17956.713203768
               LastTransitionTime: 2022-10-26 01:29:58.652331458 +0000 UTC m=+30.440319565
  ScaleDown:   NoCandidates (candidates=0)
               LastProbeTime:      2022-10-26 06:28:44.925215661 +0000 UTC m=+17956.713203768
               LastTransitionTime: 2022-10-26 01:29:58.652331458 +0000 UTC m=+30.440319565

NodeGroups:
  Name:        k3s-autoscaler
  Health:      Healthy (ready=0 unready=0 (resourceUnready=0) notStarted=0 longNotStarted=0 registered=0 longUnregistered=0 cloudProviderTarget=0 (minSize=0, maxSize=3))
               LastProbeTime:      0001-01-01 00:00:00 +0000 UTC
               LastTransitionTime: 0001-01-01 00:00:00 +0000 UTC
  ScaleUp:     NoActivity (ready=0 cloudProviderTarget=0)
               LastProbeTime:      0001-01-01 00:00:00 +0000 UTC
               LastTransitionTime: 0001-01-01 00:00:00 +0000 UTC
  ScaleDown:   NoCandidates (candidates=0)
               LastProbeTime:      2022-10-26 06:28:44.925215661 +0000 UTC m=+17956.713203768
               LastTransitionTime: 2022-10-26 01:29:58.652331458 +0000 UTC m=+30.440319565

  Name:        draining-node-pool
  Health:      Healthy (ready=0 unready=0 (resourceUnready=0) notStarted=0 longNotStarted=0 registered=0 longUnregistered=0 cloudProviderTarget=0 (minSize=0, maxSize=0))
               LastProbeTime:      0001-01-01 00:00:00 +0000 UTC
               LastTransitionTime: 0001-01-01 00:00:00 +0000 UTC
  ScaleUp:     NoActivity (ready=0 cloudProviderTarget=0)
               LastProbeTime:      0001-01-01 00:00:00 +0000 UTC
               LastTransitionTime: 0001-01-01 00:00:00 +0000 UTC
  ScaleDown:   NoCandidates (candidates=0)
               LastProbeTime:      2022-10-26 06:28:44.925215661 +0000 UTC m=+17956.713203768
               LastTransitionTime: 2022-10-26 01:29:58.652331458 +0000 UTC m=+30.440319565



BinaryData
====

**Reply by Manikanta-Reddy-Pasala** [reply]
i have only enabled metrics_server. no other stack(Prometheus, grafana) is installed.

**Reply by ifeulner** [reply]
The status output does not really fit to what you describe. 

There are four nodes in the cluster `Health: Healthy (ready=4 unready=0 (resourceUnready=0) notStarted=0 longNotStarted=0 registered=4 longUnregistered=0)`

And cluster-autoscaler sees no need to do anything - `Name: k3s-autoscaler
Health: Healthy (ready=0 unready=0 (resourceUnready=0) notStarted=0 longNotStarted=0 registered=0 longUnregistered=0 cloudProviderTarget=0 (minSize=0, maxSize=3))`

So the load does not need any scaling... you can set `min_nodes`to 1 to force autoscaler to create at least 1 node.

**Reply by captnCC** [reply]
Like @ifeulner said: from the logs, the autoscaler seems to be configured correctly.
So I think it's more a problem of the pod resource allocation. Do you have resource requirements set for them?
This is required as its the metric by which the cluster-autoscaler works (not the real machine load!) 

I have tested a similar case, just instead of starting with `1.5`, I started with a `1.6` cluster and no configured autoscaler,  then added the autoscaler pool.

**Reply by Manikanta-Reddy-Pasala** [reply]
Thank you so much for the reply. i will try adding resource limits and share you the results soon.

**Reply by Manikanta-Reddy-Pasala** [reply]
Hi @captnCC ,

i have added resource limits for my deployments. 

```yaml
resources:
            limits:
              cpu: '1'
              memory: 1500Mi
            requests:
              cpu: 200m
              memory: 512Mi
```

do I need to apply HPA also or just a resource limit is fine for auto-scaling to work?
